src/main/java/org/broadinstitute/hellbender/engine/ReferenceMemorySource.java

```diff
@@ -0,0 +1,128 @@
+        throw new UnsupportedOperationException("Iteration over entire reference not yet implemented");
+    }
+
+    /**
+     * Query a specific interval on this reference, and get back an iterator over the bases spanning that interval.
+     *
+     * See the BaseUtils class for guidance on how to work with bases in this format.
+     *
+     * @param interval query interval
+     * @return iterator over the bases spanning the query interval
+     */
+    @Override
+    public Iterator<Byte> query( final SimpleInterval interval ) {
+
+        int startIndex = (interval.getStart() - bases.getInterval().getStart());
+        // compute the number of base we're reading (+1 because those are closed intervals),
+        // then add however many bases we're skipping at the front
+        int stopIndex = (interval.getEnd() - interval.getStart() + 1) + startIndex;
+
+        return new ByteArrayIterator(bases.getBases(), startIndex, stopIndex);
+    }
+
+    /**
+     * Query a specific interval on this reference, and get back all bases spanning that interval at once.
+     * Call getBases() on the returned ReferenceSequence to get the actual reference bases. See the BaseUtils
+     * class for guidance on how to work with bases in this format.
+     *
+     * @param interval query interval
+     * @return a ReferenceSequence containing all bases spanning the query interval, prefetched
+     */
+    @Override
+    public ReferenceSequence queryAndPrefetch( final SimpleInterval interval ) {
+        return queryAndPrefetch(interval.getContig(), interval.getStart(), interval.getEnd());
+    }
+
+    /**
+     * Query a specific interval on this reference, and get back all bases spanning that interval at once.
+     * Call getBases() on the returned ReferenceSequence to get the actual reference bases. See the BaseUtils
+     * class for guidance on how to work with bases in this format.
+     *
+     * @param contig query interval contig
+     * @param start query interval start
+     * @param stop query interval stop (included)
+     * @return a ReferenceSequence containing all bases spanning the query interval, prefetched
+     */
+    @Override
+    public ReferenceSequence queryAndPrefetch( final String contig, final long start , final long stop) {
+        final int contigIndex = sequenceDictionary.getSequenceIndex(contig);
+        int startIndex = (int)(start - bases.getInterval().getStart());
+        int length = (int)(stop - start + 1);
+        byte[] basesBytes = bases.getBases();
+        if (startIndex==0 && length==basesBytes.length) {
+            // special case: no need to make a copy
+            return new ReferenceSequence(contig, contigIndex, basesBytes);
+        }
+        if (startIndex<0) {
+            throw new IllegalArgumentException("Asking for start "+start+" on contig "+contig+" but the ReferenceData only has data starting at "+bases.getInterval().getStart());
+        }
+        if (length>basesBytes.length) {
```

Comment:
This should be `if ( startIndex + length > basesBytes.length )`, not `if (length>basesBytes.length)`